File: 0503-next-greater-element-ii/0503-next-greater-element-ii.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ans = [-1] * n
        for i in range(n): 
            for j in range(i+1,i+n):
                if nums[i] < nums[j%n]:
                    ans[i] = nums[j%n]
                    break
        return ans

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        Stack<Integer> st = new Stack<>();
        int [] nge = new int[n];
        for (int i=2*n-1 ; i>=0 ; i--){
            int curr_ele = nums[i%n];
            while(!st.isEmpty() && st.peek()<= curr_ele){
                st.pop();
            }
            if (i<=n-1){
                nge[i] = st.isEmpty() ? -1 : st.peek();
            }
            st.push(curr_ele);
        }
        return nge;
    }
}